x/aswap/handler.go

```diff
@@ -0,0 +1,292 @@
+// all handlers in this package
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, cashctrl cash.Controller) {
+	bucket := NewBucket()
+
+	r.Handle(pathCreateSwap, migration.SchemaMigratingHandler("aswap", CreateSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReleaseSwap, migration.SchemaMigratingHandler("aswap", ReleaseSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReturnReturn, migration.SchemaMigratingHandler("aswap", ReturnSwapHandler{auth, bucket, cashctrl}))
+}
+
+// RegisterQuery will register this bucket as "/aswaps"
+func RegisterQuery(qr weave.QueryRouter) {
+	NewBucket().Register("aswaps", qr)
+}
+
+//---- create
+
+// CreateSwapHandler creates a swap
+type CreateSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.CoinMover
+}
+
+var _ weave.Handler = CreateSwapHandler{}
+
+// Check does the validation and sets the cost of the transaction
+func (h CreateSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &weave.CheckResult{
+		GasAllocated: createSwapCost,
+	}
+	return res, nil
+}
+
+// Deliver moves the tokens from sender to the swap account if all conditions are met.
+func (h CreateSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// apply a default for sender
+	sender := msg.Src
+	if sender == nil {
+		sender = x.MainSigner(ctx, h.auth).Address()
+	}
+
+	// create an swap object
+	swap := &Swap{
+		Metadata:     &weave.Metadata{},
+		Src:          sender,
+		PreimageHash: msg.PreimageHash,
+		Recipient:    msg.Recipient,
+		Timeout:      msg.Timeout,
+		Memo:         msg.Memo,
+	}
+	obj, err := h.bucket.Build(db, swap)
+	if err != nil {
+		return nil, err
+	}
+
+	// deposit amounts
+	senderAddr := swap.Src
+	if err := moveCoins(db, h.bank, senderAddr, weave.NewAddress(swap.PreimageHash), msg.Amount); err != nil {
+		return nil, err
+	}
+	// return id of swap to use in future calls
+	res := &weave.DeliverResult{
+		Data: obj.Key(),
+	}
+	return res, h.bucket.Save(db, obj)
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h CreateSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreateSwapMsg, error) {
+	var msg CreateSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	if IsExpired(ctx, msg.Timeout) {
+		return nil, errors.Wrap(errors.ErrInvalidInput, "timeout in the past")
+	}
+
+	// Sender must authorize this
+	if !h.auth.HasAddress(ctx, msg.Src) {
+		return nil, errors.ErrUnauthorized
+	}
+
+	// Leave the most expensive operation till we've sanity-checked everything else.
+	_, err := loadSwap(h.bucket, db, msg.PreimageHash)
+	switch {
+	case !errors.ErrEmpty.Is(err):
+		return nil, err
+	case err == nil:
```

Comment:
I think this is a dead case. `!errors.ErrEmpty.Is(nil)` is always `true` so the `err == nil` will never be called. I think you should first check for `err` being a `nil` value and then assert error type.

Reply:
Yep, missed that. Thanks!